### PR TITLE
Let Quantity accept Array-API compatible objects

### PR DESCRIPTION
**Pull request by MridulS**
This is a hack to make the code block in https://github.com/astropy/astropy/issues/15071 work and I just wanted to test this out and see how things need to work/refactor inside `Quantity` to work. The changes here are minimal and it just casts any array-api compatible object to numpy array (as suggested in https://github.com/astropy/astropy/issues/15071#issuecomment-1638031396), like a `torch` tensor or `cupy` array will be able to use astropy to create a `Quantity` object.

The following code would work with this change.
``` python
>>> import torch
>>> import astropy.units as u
>>> array = torch.asarray([1, 2, 3])
>>> u.Quantity(array)
<Quantity [1., 2., 3.]>
```
The quantity value in the above case loses the torch tensor as it's wrapped by a numpy ndarray.
``` python
>>> u.Quantity(array).value == array
False
```
Would there be interest in refactoring `Quantity` and making it work natively with array-api compatible types?

### CODE REVIEW

File: astropy/units/quantity.py
Comment (pllim) [inline astropy/units/quantity.py:463]: I think this should use the same optional dependency check as those in https://github.com/astropy/astropy/blob/main/astropy/utils/compat/optional_deps.py

File: astropy/units/tests/test_quantity_array_methods.py
Comment (pllim) [inline astropy/units/tests/test_quantity_array_methods.py:649]: Should avoid global filterwarnings like this since this can affect other tests in the test suite. 

1. Why is this warning being ignored?
2. Can it be localized with a context manager on just the affected code?

File: astropy/units/tests/test_quantity_array_methods.py
Comment (pllim) [inline astropy/units/tests/test_quantity_array_methods.py:656]: We have `astropy.tests.helper.assert_quantity_allclose()` function for such checks.

**Review by pllim — COMMENTED** [review]
Thanks! Left a general review.

I think this also needs a change log fragment but I cannot tell if it is `feature` or `api`, so I'll let `units` maintainers decide.

**Comment by pllim** [comment]
Also the failure looks related: `TypeError: object of type 'Array' has no len()`

File: astropy/units/tests/test_quantity_array_methods.py
Comment (MridulS) [reply astropy/units/tests/test_quantity_array_methods.py:649]: > Why is this warning being ignored?

importing array_api emits a warning about this being experimental.

>  Can it be localized with a context manager on just the affected code?

I tried using the pytest filterwarning fixture but I think it only works for `test_` functions?

I can make this whole class into a single functional test too, just wanted to follow the rest of the test suite in this file.

Comment (pllim) [reply astropy/units/tests/test_quantity_array_methods.py:649]: I think applying `pytest.mark.filterwarnings` on the test class might work (not 100% sure).

File: astropy/units/tests/test_quantity_array_methods.py
Comment (pllim) [inline astropy/units/tests/test_quantity_array_methods.py:647]: Do we also want to match the "experimental" warning message, so we don't accidentally ignore an unexpected UserWarning that is unrelated?

File: astropy/units/quantity.py
Comment (mhvk) [inline astropy/units/quantity.py:462]: I don't think we need this explicit test - this branch just tests for string and sequences of Quantity, and otherwise falls through to an `np.array` call which will already do the right thing. See main comment.

File: astropy/units/tests/test_quantity_array_methods.py
Comment (mhvk) [inline astropy/units/tests/test_quantity_array_methods.py:648]: I think can use a numpy version check rather than a new optional dependency.

File: astropy/units/tests/test_quantity_array_methods.py
Comment (mhvk) [inline astropy/units/tests/test_quantity_array_methods.py:650]: Maybe call it different to avoid confusion. Or `from numpy.array_api import asarray`

File: astropy/units/tests/test_quantity_array_methods.py
Comment (mhvk) [inline astropy/units/tests/test_quantity_array_methods.py:654]: I think here we need to check more directly:
```
assert type(quantity_array) is u.Quantity
assert_array_equal(quantity_array, [1, 2, 3] * u.m)
```

File: astropy/utils/compat/optional_deps.py
Comment (mhvk) [inline astropy/utils/compat/optional_deps.py:36]: I think with my suggestions, it is not needed to add an optional dependency.

File: pyproject.toml
Comment (mhvk) [inline pyproject.toml:94]: As above.

**Review by mhvk — COMMENTED** [review]
Thanks for the PR! Rather than checking for array API, I think it is better to avoid the failure that currently occurs, which can be done by replacing 
```
if isiterable(value) and len(value) > 1
```
with
```
if isinstance(value, (list, tuple)) and len(value) > 1
```

I think this is actually a bit better than `isiterable` anyway, though it is an API change so will need a corresponding changelog entry...

**Comment by MridulS** [comment]
Thanks for the reviews! Now I just avoid the iterable check as suggested, this should fix https://github.com/astropy/astropy/issues/15071 but we would need `array_api_compat `if astropy wants to natively support other array types (currently it will cast them using ndarray). I'm not sure with the API changelog I have written so feel free to overwrite it :D

**Review by mhvk — APPROVED** [review]
OK, looks good! I rephrased the API changelog entry a bit, hopefully making it (even) clearer, and also squashed the commits. Approving now and setting up for auto-merge...

**Pull request merged**
